xarray/core/coordinates.py

```diff
@@ -23,49 +34,58 @@
 _THIS_ARRAY = ReprObject("<this-array>")
 
 
-class AbstractCoordinates(collections.abc.Mapping):
-    def __getitem__(self, key):
-        raise NotImplementedError
+class AbstractCoordinates(Mapping[Hashable, "DataArray"]):
+    _data = None  # type: Union["DataArray", "Dataset"]
 
-    def __setitem__(self, key, value):
+    def __getitem__(self, key: Hashable) -> "DataArray":
+        raise NotImplementedError()
+
+    def __setitem__(self, key: Hashable, value: Any) -> None:
         self.update({key: value})
 
     @property
-    def indexes(self):
+    def _names(self) -> Set[Hashable]:
```

Comment:
Is this new? Should this be on the abstract class?

Reply:
It's a property invoked by the methods of the abstract class, defined in both the implementations.

Reply:
Here, I'm not so sure - do all concrete classes need to define this? Generally we wouldn't define private methods on an abstract class (even if all currently concrete classes implement it), because the implementation would be up to the concrete class?

Reply:
What I basically did was turn AbstractCoordinates from a mixin (that is, a class that invokes own methods that are only defined in a subclass) into a C++-style abstract class.
mypy, justifiably, doesn't like mixins - short of hacking your way through with ``# type`` tags.

This, in synthesis, is how it was before:
```
class C:
    def g(self) -> int:
        return self.f()


class D(C):
    def f(self) -> int:
        return 123
```
mypy correctly states:
```3: error: "C" has no attribute "f"```

Reply:
Right - I agree this should either be an abstract class without implementations or a mixin with implementations - not a bit of each. It's a bit odd we call this class `Abstract` but then enforce a `_names` property.

That a nice upside of mypy - we get to find those things out

I agree we're in a reasonable local maximum here.

Reply:
Yes, thanks for fixing this up!